twilio/base/client_base.py

```diff
@@ -85,8 +88,19 @@ def request(
 
         :returns: Response from the Twilio API
         """
-        auth = self.get_auth(auth)
+
         headers = self.get_headers(method, headers)
+
+        ##If credential provider is provided by user, get the associated auth strategy
+        ##Using the auth strategy, fetch the auth string and set it to authorization header
+        auth_strategy = None ##Initialization
+        if self.credential_provider:
+            auth_strategy = self.credential_provider.to_auth_strategy()
```

Comment:
If auth strategy exists, reuse it and token should not be fetched every single time
should be fetched only once it expires

Reply:
auth strategy should be created only once - test this case

Reply:
done